### PR TITLE
doc: document release criteria

### PR DESCRIPTION
**Pull request by nashif**
Overhaul of release criteria and bug labeling. Introducing severities
and blocker bugs as well as defining board tiers.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

### CODE REVIEW

File: doc/project/release_process.rst
Comment (dleach02) [inline doc/project/release_process.rst:186]: My primary problem with having High, Medium, Low and Blocker status is what is the point then of the High/Medium/Low? To me, if it is a blocker bug then it is a High bug. It has to be fixed. We used to not allow a release to occur if a High bug occurred but now we do because of this concept of a blocker. So the effect is that a Blocker is a "higher level" bug. If I would apply a numeric to this then it would be

1 - Blocker
2 - High
3 - Medium
4 - Low

Effectively 4 levels/classifications of bugs. Now I know you are really using "blocker" as an overlay modifier but I don't see the point. It is a class 1 bug that must be fixed or the release can't happen.

**Review by hakehuang — DISMISSED** [review]
just suggestion to add the definition with a customer anchor, so that we can be referenced inside.

File: doc/project/release_process.rst
Comment (kartben) [inline doc/project/release_process.rst:156]: I suggest using a table here for improved readability

```suggestion
+----------+------------------------------------------------------------------+
| Severity | Description                                                      |
+==========+==================================================================+
| Blocker  | Such an error makes it impossible to proceed with using or       |
|          | testing the software. See below for more details.                |
+----------+------------------------------------------------------------------+
| Critical | Bug results in a crash of the system or one or more components   |
|          | of the system. If the failed function is unusable and there is   |
|          | no acceptable alternative method to achieve the advertised       |
|          | desired outcome then the severity will be stated as critical.    |
+----------+------------------------------------------------------------------+
| Major    | As with critical, but an acceptable alternative exists.          |
+----------+------------------------------------------------------------------+
| Moderate | Causes the system to produce incorrect, incomplete, or           |
|          | inconsistent results.                                            |
+----------+------------------------------------------------------------------+
| Minor    | No crashes or damage to usability of the system and desired      |
|          | results can be easily obtained by working around the defects.    |
+----------+------------------------------------------------------------------+
| Cosmetic | Related to look and feel of the system.                          |
+----------+------------------------------------------------------------------+
```

File: doc/project/release_process.rst
Comment (kartben) [inline doc/project/release_process.rst:162]: nit: extra blank line

File: doc/project/release_process.rst
Comment (kartben) [inline doc/project/release_process.rst:92]: formatting seems inconsistent across the file (max line length)

File: doc/project/release_process.rst
Comment (kartben) [inline doc/project/release_process.rst:174]: I would swap it around (we are in the section talking about blockers, not the section about what milestones and RC3 mean, if that makes sense)

```suggestion
A bug that is granted `blocker` status can be merged to 'main' and included in the release all the way until the hard freeze (RC3) milestone.
```

File: doc/project/release_process.rst
Comment (kartben) [inline doc/project/release_process.rst:104]: i would be tempted to s/bug/issue/ across mostly the whole doc, except for the instances where we do mean "bug". as in a defect in code?
But not sure that will work with e.g. the current definition of blocker below. 
Just thinking about the "blocker" issue we had for 3.3 where we didn't want to ship until the issue with web analytics and cookie consent would be addressed.

File: doc/project/release_process.rst
Comment (kartben) [inline doc/project/release_process.rst:197]:
```suggestion
- Bugs reported against platforms of this tier are to be evaluated and treated as
```

File: doc/project/release_process.rst
Comment (kartben) [inline doc/project/release_process.rst:209]:
```suggestion
- Bugs reported against platforms of this tier are to be evaluated and treated as
```

File: doc/project/release_process.rst
Comment (kartben) [inline doc/project/release_process.rst:203]:
```suggestion
- Commitment from a specific team to run tests using twister device
```

File: doc/project/release_process.rst
Comment (kartben) [inline doc/project/release_process.rst:113]: would be worth expanding on what this means? Would adding "... (CI builds pass)" be enough maybe?

File: doc/project/release_process.rst
Comment (kartben) [reply doc/project/release_process.rst:109]: also maybe misses a verb? "Footprint/Performance Tracking data is updated and available at XXX (no specific KPIs)"?

File: doc/project/release_process.rst
Comment (kartben) [inline doc/project/release_process.rst:111]: also maybe missing a verb

```suggestion
- Documentation is complete and Release Notes are up-to-date.
```

File: doc/project/release_process.rst
Comment (kartben) [inline doc/project/release_process.rst:112]: I would think that it's already implicitly assumed that the release criteria apply to all subsystems, so maybe drop it?

File: doc/project/release_process.rst
Comment (kartben) [inline doc/project/release_process.rst:119]: no need to make a reference to the past imo

```suggestion
Priority of bugs is not enough to make release decisions: not all critical bugs are of highest
```

**Review by kartben — DISMISSED** [review]

**Comment by nashif** [comment]
> @nashif, when you were describing the way coverity classifies issues found in the release readiness meeting you used the term "impact" which got me thinking. Maybe we should consider using the same term when classifying issues instead of "priority". This might be why I get so caught up with the general terminology of when you define an additional property of "blocker", which to me is just priority 0 (if 1, 2, 3 are high, medium, low).

we use `severity` in this PR, which is synonym to `Impact`. I am fine using Impact instead of severity, priority is something else, and a blocker is neither as we have already discussed somewhere else.

File: doc/project/release_process.rst
Comment (dleach02) [inline doc/project/release_process.rst:135]: Should we just remove priority from our documentation and just document severity/impact with the additional modifier of "blocker" for issues that we deem needing a fix before release.

File: doc/project/release_process.rst
Comment (dleach02) [inline doc/project/release_process.rst:207]: Highest priority should they have the "blocker" modifier?

Comment (nashif) [reply doc/project/release_process.rst:207]: they could be a blocker if there is a justification, I would not say they are automatically blockers.

File: doc/project/release_process.rst
Comment (nashif) [reply doc/project/release_process.rst:135]: Based on the discussions we had, priority can be used by teams and maintainers.  We should leave this at least for now and see how things will work out and can be removed later.

File: doc/project/release_process.rst
Comment (nashif) [reply doc/project/release_process.rst:205]: > but this is where I get mentally caught up with "priority" "severity" "impact" "blocker". Too many terms with too many implied meanings

not sure why you are putting impact in here as if it is something else we have to track, this is just the same as severity, I mentioned that in the context of coverity, but Impact=Severity. 

Blocker is used to mark any bug or issue of any severity as something that needs to be addressed before a release, everything else can be addressed later and should not block the release. 

Some critical bug in an optional subsystem might not be considered as something that should block a release, if it should, it will get the blocker label, but not all critical bugs are blockers.

File: doc/project/release_process.rst
Comment (cfriedt) [inline doc/project/release_process.rst:413]: I've had some feedback that says that the support period should actually be longer. The main reason being that it often takes 2 years just to get a product to market, and some product makers want longer-term stability.

On the other hand, Nordic, who bases the NRF Connect SDK on Zephyr needs the stability but also needs the feature updates on a regular basis. It's a bit of a double-edged sword.

Comment (nashif) [reply doc/project/release_process.rst:413]: this is orthogonal, the LTS timeline and cadence it s a different topic to be changed sperately. I am not changing this text in this PR.

**Review by jhedberg — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

File: doc/project/release_process.rst
Comment (fabiobaltieri) [inline doc/project/release_process.rst:164]: Feels a bit too granular to me, but guess we can go with it and see how it works out. Bet we'll only end up using few of these.

File: doc/project/release_process.rst
Comment (fabiobaltieri) [reply doc/project/release_process.rst:193]: I'm with David on this, not clear what we'll do with all the categorization, ultimately it's a binary decision on whether to release or not.

File: doc/project/release_process.rst
Comment (fabiobaltieri) [inline doc/project/release_process.rst:214]: Should we give an indication of expected time-to-resolution? Just to give a guideline of how a support team should be staffed. Commitment here means from the platform vendor, right?

**Comment by fabiobaltieri** [comment]
(tagging DNM since it's flagged as WIP in the comment)

File: doc/project/release_process.rst
Comment (dkalowsk) [inline doc/project/release_process.rst:93]: "for a public release to ship."

File: doc/project/release_process.rst
Comment (dkalowsk) [reply doc/project/release_process.rst:164]: I'll agree with @fabiobaltieri  here.  Not really sure there is a reason to have a difference between `Minor` and `Cosmetic` The rest of the classifications look correct to me.

File: doc/project/release_process.rst
Comment (dkalowsk) [inline doc/project/release_process.rst:188]: does the term "Zephyr" here include libraries?  To mean that reads more on the specific core kernel library than the whole ecosystem.  Just looking for clarification if, for example, some critical issue is found with LittleFS (or another module library) would that fall under this umbrella?

File: doc/project/release_process.rst
Comment (dkalowsk) [reply doc/project/release_process.rst:193]: I view it as an indicator to external parties to use when making an assessment on the state of the release and if they will want to upgrade to it.  The project likely will only use a binary setting, but external entities may need more detail to make a decision on an upgrade.  For example, upgrading to a release there may exist a security or safety requirement that requests we make a detailed audit of the "High" issues.  The project isn't using them, but the snapshot of state at release is something many groups use for determining upgrades.

File: doc/project/release_process.rst
Comment (dkalowsk) [reply doc/project/release_process.rst:205]: These lines on the Tier criteria setting a priority also confuse me.  Especially since line 134 of this document states:

> Bug priorities can be used by owners of bugs in an area or subsystem to manage
their workflow and priorities.

But here (and in the lower Tier definitions) the document is specifically stating a priority for the bug/issue filed.  Is s/priority/severity needed here?

**Comment by nashif** [comment]
> These lines on the Tier criteria setting a priority also confuse me. Especially since line 134 of this document states:
> 
> Bug priorities can be used by owners of bugs in an area or subsystem to manage
their workflow and priorities.

> But here (and in the lower Tier definitions) the document is specifically stating a priority for the bug/issue filed. Is s/priority/severity needed here?


We should be able to use the term priority without having a label for it, to avoid any confusion maybe use urgency?

File: doc/project/release_process.rst
Comment (nashif) [reply doc/project/release_process.rst:203]: We should be able to use the term priority without having a label for it, to avoid any confusion maybe use urgency?

**Review by aescolar — COMMENTED** [review]
> "Severity shall be set based on the context of the bug and not in relation to the
overall impact on the codebase."

It is possible to define things like this, but we need to be careful: There is always going to be

* an "overall importance a bug has for the whole project",
* an impact a bug has on a subsystem / feature / platform
* the importance of that subsytem / feature / platform.

The overall priority/severity/drive to fix an issue will be the result of the `{importance of the affected features}*{effect the bug has on the features}*{importance of the affected platforms}`

If we allow general labels like "severity" or "priority" to be applied both to the overall importance of the bug, and to the importance that is has on a given feature, we are going to end up with confused reporting and triaging.

Anyhow, I think first we need to agree a basic terminology, and be consistent with it, because right now this text seems to be using its own terms with different meanings.

Sidenotes: The board tiering makes overall sense to me.

File: doc/project/release_process.rst
Comment (nashif) [reply doc/project/release_process.rst:212]: > Commitment here means from the platform vendor, right?

yes.

**Comment by dleach02** [comment]
> Anyhow, I think first we need to agree a basic terminology, and be consistent with it, because right now this text seems to be using its own terms with different meanings.

Agreed. Terms first then consistent usage with rest of document

**Comment by nashif** [comment]
> Anyhow, I think first we need to agree a basic terminology, and be consistent with it, because right now this text seems to be using its own terms with different meanings.

what own terms with what different meanings, can you please elaborate? This text is trying to solve a problem we have right now. Severity of a bug should not be set based on who uses the feature and how popular a feature is, it needs to apply to the  context of where the bug. This is the only way we will have a general view of what systems and platforms need attention and have problems. If we set severity based on usage and poplarity, then we end up with an unbalanced picture where most bugs are marked as minor (the same way we have those marked with low priority right now). 
Setting severity of a bug will also become a function of who sets the label and how much they care about a certain subsystem.

File: doc/project/release_process.rst
Comment (nashif) [reply doc/project/release_process.rst:199]: > Shouldn't introducing any bug (for any platform) in a PR be a valid reason to block a PR?

yes, this means a PR is blocked by CI in cases there are failures in those platforms, which we run in CI.

> The fact that we may only typically detect issues in tier 0 platforms in CI means we will normally only realize about those right away, but if somebody catches another bug by eye, or thru other tests that would still be a valid reason to reject it.

correct, that  is when you block a PR through review and requesting changes. The above is about blocking the PR in CI.

> Maybe this should say something in the lines of "Tests on tier 0 platforms are run per PR on CI, and therefore are the ones in which possible bugs will be detected earliest"

added that.

File: doc/project/release_process.rst
Comment (nashif) [reply doc/project/release_process.rst:202]: if issues on tier 0 platforms are not dealt with with the highest priority, development is stalled. I am most talking about post merge issues, not something in a PR which will be dealt with as a any other change and can just be rejected.

**Comment by aescolar** [comment]
>  what own terms with what different meanings, can you please elaborate?

The difference of meaning would seem to be in statements like:
"Severity shall be used to classify bugs based on impact to users and community in general."
"Severity: The extent to which an issue or defect can affect the software and defines the
impact that a given defect has on the system."  (i.e. overall severity to the project)
"Severity shall be set based on the context of the bug and not in relation to the
overall impact on the codebase." 

> This text is trying to solve a problem we have right now. Severity of a bug should not be set based on who uses the feature and how popular a feature is, it needs to apply to the context of where the bug. This is the only way we will have a general view of what systems and platforms need attention and have problems. 

Should we, then, try to be more explicit with what we mean and call it "impact on the feature" instead of "severity". (or whatever other wording you'd prefer)

> If we set severity based on usage and popularity, then we end up with an unbalanced picture where most bugs are marked as minor 

We should also reflect with some label during triagging what is the overall priority of the bug for the project as it is very relevant. We could keep using "priority" for those.
Say a "minor" bug that affects all platforms and apps even if not critically, may warrant more attention than, say, a crash that only triggers in some not-common platform when an obscure combination of features is used.

I'd state in the documentation something conceptually like:
"Each bug has an _overall project priority_, which we quantize as high,medium or low. This _overall project priority_ should be estimated as the bug _impact on the feature_ weighted by the _importance of the affected feature(s)_ and by the  _importance of the affected platform(s)_."
"This bug _overall project priority_ is estimated during bug triaging, and guides the overall focus and priority the project will have to fix the bug, as well as its impact on the release criteria"

**Comment by MaureenHelm** [comment]
@nashif ping

**Pull request closed without merging**
